\markdownRendererHeadingOne{Related Work}\markdownRendererInterblockSeparator
{}With the increasing digitalisation of research processes, there has been a significant call for the wider adoption of interoperable sharing of data and its associated metadata. For a comprehensive overview and recommendations, in particular for data, we refer to \markdownRendererCite{1}+{}{}{doi:10.1016/j.patter.2020.100136}, which highlights the wide variety of metadata and documentation that the literature prescribes for enabling data reuse. \markdownRendererInterblockSeparator
{}Here we focus on approaches for bundling research artefacts along with their metadata. This notion has a long history behind it \markdownRendererCite{1}+{}{}{doi:10.1190/1.1822162}, but recent approaches have followed three main strands: 1) publishing to centralised repositories; 2) packaging approaches similar to RO-Crate; and 3) bundling the computational workflow around a scientific experiment. \markdownRendererInterblockSeparator
{}% ## Repositories % % * DataCite Metadata \markdownRendererCite{1}+{}{}{doi:10.14454/3w3z-sa82} % * Open Science Framework\markdownRendererInterblockSeparator
{}\markdownRendererHeadingTwo{Bundling and Packaging Digital Research Artefacts}\markdownRendererInterblockSeparator
{}The challenge of describing computational workflows was one of the main motivations for the early proposal of \markdownRendererEmphasis{Research Objects} (RO) \markdownRendererCite{1}+{}{}{doi:10.1016/j.future.2011.08.004} as first-class citizens for sharing and publishing. The RO approach involves bundling datasets, workflows, scripts and results along with traditional dissemination materials like journal articles and presentations, forming a single package. Crucially, these resources are not just gathered, but also individually typed, described and related to each other using semantic vocabularies. As pointed out in \markdownRendererCite{1}+{}{}{doi:10.1016/j.future.2011.08.004} an open-ended \markdownRendererEmphasis{Linked Data} approach is not sufficient for scholarly communication: a common data model is also needed in addition to common and best practices for managing and annotating lifecycle, ownership, versioning and attributions.\markdownRendererInterblockSeparator
{}Considering the FAIR principles, we can say with hindsight that the initial RO approaches strongly targeted \markdownRendererEmphasis{Interoperability}, with a particular focus on the reproducibility of \markdownRendererEmphasis{in-silico experiments} involving computational workflows and the reuse of existing RDF vocabularies. \markdownRendererInterblockSeparator
{}The first implementation of Research Objects for sharing workflows in myExperiment \markdownRendererCite{1}+{}{}{doi:10.1093/nar/gkq429} was based on RDF ontologies \markdownRendererCite{1}+{}{}{newman2009}, building on Dublin Core, FOAF, SIOC, Creative Commons and OAI-ORE to form myExperiment ontologies for describing social networking, attribution and credit, annotations, aggregation packs, experiments, view statistics, contributions, and workflow components \markdownRendererCite{1}+{}{}{myExperimentOntology2009}.\markdownRendererInterblockSeparator
{}% * RO-Bag-It % * BDBag % * Ontologies ( OAI-ORE, OAC/AO/OA)\markdownRendererInterblockSeparator
{}\markdownRendererHeadingTwo{FAIR Digital Objects}\markdownRendererInterblockSeparator
{}FAIR Digital Objects (FDO) \markdownRendererCite{1}+{}{}{doi:10.3390/publications8020021} have been proposed as a conceptual framework for making digital resources available in a Digital Objects (DO) architecture that encourages active use of the objects and their metadata. In particular, an FDO has five parts: (i) The FDO \markdownRendererEmphasis{content}, bit sequences stored in an accessible repository; (ii) a \markdownRendererEmphasis{Persistent Identifier} (PID) such as a DOI that identifies the FDO and can resolve these parts; (iii) Associated rich \markdownRendererEmphasis{metadata}, as separate FDOs; (iv) Type definitions, also separate FDOs; (v) Associated \markdownRendererEmphasis{operations} for the given types. A Digital Object typed as a Collection aggregates other DOs by reference.\markdownRendererInterblockSeparator
{}As an "\markdownRendererLink{abstract protocol}{https://www.dona.net/sites/default/files/2018-11/DOIPv2Spec_1.pdf}{https://www.dona.net/sites/default/files/2018-11/DOIPv2Spec_1.pdf}{}", DOs could be implemented in multiple ways. One suggested implementation is the \markdownRendererLink{FAIR Digital Object Framework}{https://fairdigitalobjectframework.org/}{https://fairdigitalobjectframework.org/}{}, based on HTTP and the Linked Data Principles. While there is agreement on using PIDs based on DOIs, consensus on how to represent common metadata, core types and collections as FDOs has not yet been reached. We argue that RO-Crate can play an important role for FDOs:\markdownRendererInterblockSeparator
{}\markdownRendererOlBeginTight
\markdownRendererOlItemWithNumber{1}By providing a predictable and extensible serialisation of structured metadata.\markdownRendererOlItemEnd 
\markdownRendererOlItemWithNumber{2}By formalising how to aggregate digital objects as collections (and adding their context).\markdownRendererOlItemEnd 
\markdownRendererOlItemWithNumber{3}By providing a natural Metadata FDO in the form of the RO-Crate Metadata File.\markdownRendererOlItemEnd 
\markdownRendererOlItemWithNumber{4}By being based on Linked Data and schema.org vocabulary, meaning that PIDs already exist for common types and properties.\markdownRendererOlItemEnd 
\markdownRendererOlEndTight \markdownRendererInterblockSeparator
{}At the same time, it is clear that the goal of FDO is broader than that of RO-Crate; namely, FDOs are active objects with distributed operations, and add further constraints such as PIDs for every element. These features improve FAIR features of digital objects and are also useful for RO-Crate, but they also severely restrict the infrastructure that needs to be implemented and maintained in order for FDOs to remain available. RO-Crate, on the other hand, is more flexible: it can minimally be used within any file system structure, or ideally exposed through a range of Web-based scenarios. A \markdownRendererEmphasis{FAIR profile of RO-Crate} (e.g. enforcing PID usage) will fit well within a FAIR Digital Object ecosystem.\markdownRendererInterblockSeparator
{}\markdownRendererHeadingTwo{Packaging Workflows}\markdownRendererInterblockSeparator
{}The use of computational workflows, typically combining a chain of open source tools in an analytical pipeline, has gained prominence, in particular in the life sciences. Workflows may have initially been used to improve computational scalability, but they also assist in making computed data results FAIR \markdownRendererCite{1}+{}{}{doi:10.1162/dint_a_00033}, for instance by improving reproducibility \markdownRendererCite{1}+{}{}{10.1016/j.future.2017.01.012}, but also because programmatic data usage help propagate their metadata and provenance \markdownRendererCite{1}+{}{}{doi:10.1002/cpe.1228}. At the same time, however, workflows raise additional FAIR challenges, since they can be considered important research artefacts themselves, posing the problem of capturing and explaining the computational methods behind the analysis they perform \markdownRendererCite{1}+{}{}{doi:10.3233/DS-190026}.\markdownRendererInterblockSeparator
{}Even when researchers follow current best practices for workflow reproducibility, \markdownRendererCite{1}+{}{}{doi:10.1016/j.cels.2018.03.014} \markdownRendererCite{1}+{}{}{doi:10.1016/j.future.2017.01.012} the communication of outcomes through traditional academic publishing routes relying on a textual representation adds barriers that hinder reproducibility and FAIR use of the knowledge previously captured in the workflow.\markdownRendererInterblockSeparator
{}As a real-life example, let us look at a metagenomics article \markdownRendererCite{1}+{}{}{doi:10.1038/s41586-019-0965-1} where the authors have gone to extraordinary efforts to document the individual tools that have been reused, including their citations, versions, settings, parameters and combinations. The \markdownRendererEmphasis{Methods} section is 2 pages in tight double-columns with 24 additional references, supported by the availability of data on an FTP server (60 GB) \markdownRendererCite{1}+{}{}{ebi_ftp_umgs2019} and of open source code in GitHub \markdownRendererLink{Finn-Lab/MGS-gut}{https://github.com/Finn-Lab/MGS-gut}{https://github.com/Finn-Lab/MGS-gut}{} \markdownRendererCite{1}+{}{}{finn-lab-mgsgut}, including the pipeline as shell scripts and associated analysis scripts in R and Python.\markdownRendererInterblockSeparator
{}This attention to reporting detail for computational workflows is unfortunately not yet the norm, and although bioinformatics journals have strong \markdownRendererEmphasis{data availability} requirements, they frequently do not require authors to include or cite \markdownRendererEmphasis{software, scripts and pipelines} used for analysing and producing results \markdownRendererCite{1}+{}{}{soilandreyes_tweet_2020} – rather, authors might be penalised for doing so [cite?] as it would detrimentally count against arbitrary limits on number of pages and references.\markdownRendererInterblockSeparator
{}However detailed this additional information might be, another researcher who wants to reuse a particular computational method may first want to assess if the described tool or workflow is Re-runnable (executable at all), Repeatable (same results for original inputs on same platform), Reproducible (same results for original inputs with different platform or newer tools) and ultimately Reusable (similar results for different input data), Repurposable (reusing parts of the method for making a new method) or Replicable (rewriting the workflow following the method description). \markdownRendererCite{1}+{}{}{doi:10.3389/fninf.2017.00069}\markdownRendererCite{1}+{}{}{goble_presentation_2016}\markdownRendererInterblockSeparator
{}Following the textual description alone, researchers would be forced to jump straight to evaluate "Replicable" by rewriting the pipeline from scratch. This can be expensive and error-prone. They would firstly need to install all the software dependencies and download reference datasets. This can be a daunting task in and of itself, which may have to be repeated multiple times as workflows typically are developed at small scale on desktop computers, scaled up to local clusters, and potentially put into production using cloud instances, each of which will have different requirements for software installations.\markdownRendererInterblockSeparator
{}In recent years the situation has been greatly improved by software packaging and container technologies like Docker and Conda, which have seen increased adoption in life sciences \markdownRendererCite{1}+{}{}{doi:10.1007/s41019-017-0050-4} thanks to collaborative efforts such as BioConda \markdownRendererCite{1}+{}{}{doi:10.1038/s41592-018-0046-7} and BioContainers \markdownRendererCite{1}+{}{}{doi:10.1093/bioinformatics/btx192}, and support by Linux distributions (e.g. Debian Med \markdownRendererCite{1}+{}{}{doi:10.1186/1471-2105-11-S12-S5}). As of May 2021, more than 7000 software packages are available \markdownRendererLink{in BioConda alone}{https://anaconda.org/bioconda/}{https://anaconda.org/bioconda/}{}, and 9000 containers \markdownRendererLink{in BioContainers}{https://biocontainers.pro/#/registry}{https://biocontainers.pro/#/registry}{}. Docker and Conda have gained integration in workflow systems such as Snakemake \markdownRendererCite{1}+{}{}{doi:10.1093/bioinformatics/bts480}, Galaxy \markdownRendererCite{1}+{}{}{doi:10.1093/nar/gky379} and Nextflow \markdownRendererCite{1}+{}{}{doi:10.1038/nbt.3820}, meaning a downloaded workflow definition can now be executed on a "blank" machine (except for the workflow engine) with the underlying analytical tools installed on demand – but even here there is a reproducibility challenge, for instance \markdownRendererLink{Docker Hub's retention policy will expire container images after 6 months}{https://www.docker.com/blog/docker-hub-image-retention-policy-delayed-and-subscription-updates/}{https://www.docker.com/blog/docker-hub-image-retention-policy-delayed-and-subscription-updates/}{}, or lack of recording versions of transitive dependencies of Conda packages could cause incompatibilities if the packages are subsequently updated. Except for brief metadata in their repositories, these containers and packages do not capture any semantic relationships of their content – rather their opaqueness and wrapping of arbitrary binary tools makes such relationships harder to find.\markdownRendererInterblockSeparator
{}From this we see that computational workflows are themselves complex digital objects that needs to be recorded not just as files, but in the context of their execution environment, dependencies and analytical purpose in research – as well as their FAIR metadata (e.g. version, license, attribution and identifiers).\relax